realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -628,7 +628,7 @@ private void completedAsyncRealmObject(QueryUpdateTask.Result result) {
                 } else if (compare > 0) {
                     // the caller has advanced we need to
                     // retry against the current version of the caller if it's still empty
-                    if (RealmObject.isValid(proxy)) { // already completed & has a valid pointer no need to re-run
+                    if (RealmObject.isLoaded(proxy)) { // already completed & has a valid pointer no need to re-run
```

Comment:
I doubt it might the root cause of #4086 , but i am not sure. This behavior has some potential problems. See https://gist.github.com/beeender/9a37c81d56326f02be734a3087da0f84#file-breaking-java-L66